### ISSUE TITLE
How to train FasterSeg with customized labels?

### ISSUE THREAD

**Issue by KHeap25** [VITA-Group/FasterSeg#46]
Hello, 
it's very interesting to train and use FasterSeg with own custom data.
To get information about that, I read the comments of this [issue description](https://github.com/VITA-Group/FasterSeg/issues/44#issuecomment-688599075). 

Based on the description linked above, I did the following steps:

1. prepare my own labels in [labels.py](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py)
2. create my own annotated pictures with [createTrainIdLabelImgs.py](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/preparation/createTrainIdLabelImgs.py)
2.1 make sure that hight and width are [divisible by 32](https://github.com/VITA-Group/FasterSeg/issues/34) 
3. prepare file name lists as "dataset files" for data loading [see examples here](https://github.com/VITA-Group/FasterSeg/tree/master/tools/datasets/cityscapes)
4. adjust the [config file](https://github.com/VITA-Group/FasterSeg/blob/master/search/config_search.py#L40) with number of classes

Are there any other points in the FasterSeg repository which I need to adjust for using customized labels?

For example:
In [cityscapes.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/cityscapes/cityscapes.py), [camvid.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/camvid/camvid.py) and [bdd.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/bdd/bdd.py) are some methodes like [get_class_colors()](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/tools/datasets/cityscapes/cityscapes.py#L11) and [get_class_names()](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/tools/datasets/cityscapes/cityscapes.py#L20) which return color or class names of the cityscapes data.

Is it neccessary to add the customized labels to this methods?
For which purpose are this methods?

It would be great if you can give me some hints to answer this questions, so I can run the [training process](https://github.com/VITA-Group/FasterSeg#11-pretrain-the-supernet) with customized labels.

Best regards

**Comment by chenwydj** [comment]
Hi @KHeap25 !

Thank you very much for your interests in our work!

First, I have to admit that I have never run this repo for a brand new dataset from scratch. What I can do is try to help you with and bug or problem.

Some comments to your steps:
1. As I remember the gound truth files in cityscapes/bdd/camvid are single-channel image files with integers. Each integer represents the class index for each pixel. That means as long as you prepare your ground truth files in the same way, you can run our code and don't have to write your own `labels.py` and `createTrainIdLabelImgs.py` (unless you want to make your dataset as standard as cityscapes)
2. During training please make the height and width of cropped patches divisible by 64. Sorry I made mistakes in responses to previous issues (I already corrected them).
3. It is recommended to implement your `get_class_names()` function, as it will be used to print the results to the terminal (see [here](https://github.com/VITA-Group/FasterSeg/blob/master/search/eval.py#L64) for usage). The function `get_class_colors()` is only used when you want to generate predictions in color (something like the front teaser images in our Readme) (see [here](https://github.com/VITA-Group/FasterSeg/blob/master/train/test.py#L61) for usage).

Hope that helps.

**Comment by KHeap25** [comment]
Thanks @chenwydj for your reply!

In relation to your recommendation above, I implement a new version of the ```get_class_colors()``` and the ```get_class_names()```  methods in [cityscapes.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/cityscapes/cityscapes.py), [camvid.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/camvid/camvid.py) and [bdd.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/bdd/bdd.py) based on the customized labels.
Furthermore there are some lines in [test.py](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/train/test.py#L25) that need to be adjust for using the custom labels.

After passing the step [pretrain the supernet](https://github.com/VITA-Group/FasterSeg#11-pretrain-the-supernet), I got the following error during the step [search the architecture](https://github.com/VITA-Group/FasterSeg#12-search-the-architecture).

![image](https://user-images.githubusercontent.com/66534155/97180945-023f3a80-179b-11eb-9b4d-1b40ea4a6d80.png)

It looks like something with the "TensorRT" latency test went wrong. 
Maybe in [darts_utils.py](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/tools/utils/darts_utils.py#L113).
Do you have an idea how this error can be fixed?

If I remove "TensorRT" from the system, I am able to run all the [training and evaluation steps](https://github.com/VITA-Group/FasterSeg/tree/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916#12-search-the-architecture).
But then, PyTorch is used for the latency tests.

![image](https://user-images.githubusercontent.com/66534155/97188840-3ec36400-17a4-11eb-9b35-213803708fdf.png)

Are there any important differences between TensorRT and PyTorch for the latency test?

I look forward to hearnig from you. 

Kind regards

**Comment by chenwydj** [comment]
I would guess the tensorrt meets some problem, although you have installed it.

Here is the place where the function using tensorrt is imported, you may want to comment out this part:
https://github.com/VITA-Group/FasterSeg/blob/master/search/operations.py#L25

**Comment by KHeap25** [comment]
Hi @chenwydj,

firstly let me summarize the points which I adjusted for using own customized labels.

* prepare the own custom labels in [labels.py](https://github.com/mcordts/cityscapesScripts/blob/41f4ecbd62d151a1392573ff785a02c56c8c45ef/cityscapesscripts/helpers/labels.py#L62)
* adapt the [get_class_colors()](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/tools/datasets/cityscapes/cityscapes.py#L11) and [get_class_names()](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/tools/datasets/cityscapes/cityscapes.py#L20) methods in [cityscapes.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/cityscapes/cityscapes.py), [camvid.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/camvid/camvid.py) and [bdd.py](https://github.com/VITA-Group/FasterSeg/blob/master/tools/datasets/bdd/bdd.py), which return color or class names of the labels.
* change the trainID2id mapping in [test.py](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/train/test.py#L25) for the customized labels
* set the number of classes in [config_search.py](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/search/config_search.py#L40) and [config_train.py](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/train/config_train.py#L42)
* the annotated pictures can be created with [createTrainIdLabelImgs.py](https://github.com/mcordts/cityscapesScripts/blob/41f4ecbd62d151a1392573ff785a02c56c8c45ef/cityscapesscripts/preparation/createTrainIdLabelImgs.py)

After doing the steps, I was able to run a training process with customized labels [(with PyTorch for latency test)](https://github.com/VITA-Group/FasterSeg/issues/46#issuecomment-716609584). 

Based on that, I have some questions about the [logger](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/train/train.py#L44) for TensorBoard monitoring.

* Are you logging the [loss](https://github.com/VITA-Group/FasterSeg/blob/f4d53ef7f4eb9a6f0f72c07d141aef3c2b796916/train/train.py#L266) and some other things like mIoU, etc. for all steps in each epoch? 
* How should I interpret the graphs from [step 1.2](https://github.com/VITA-Group/FasterSeg#12-search-the-architecture)? See the following picture for example.
![Issue](https://user-images.githubusercontent.com/66534155/99428938-98343400-2907-11eb-8871-74ca0f274f79.JPG)
* Do you use some cunstructions to identify or overcome overfitting?

It would be great if you can give me some advice, which would help me to understand the results of the training steps better.

Kind regards

**Comment by chenwydj** [comment]
Hi @KHeap25,

The "objective" indicates Eq. 5 in our Appendix B. It is a combined target of accuracy and latency, adopted from [Tan et al., 2019](https://arxiv.org/abs/1807.11626).

The other parts is FPS, and the code is [here](https://github.com/VITA-Group/FasterSeg/blob/master/search/train_search.py#L285). Arch0 indicates teacher net, arch1 the student. FPS0 indicates the architecture that aggregates outputs from [1/8, 1/32] branches, and FPS1 the aggregation from [1/16, 1/32].

Hope that helps!

**Comment by i-am-nut** [comment]
@KHeap25 can you share your code to see how your modifications got afterall?

**Comment by MouseOverG** [comment]
Hey @emersonjr,

 @KHeap25 was in the project with me. You can find our code here: https://github.com/Gaussianer/FasterSeg
